services/sonar/sonar-violations.service.js

```diff
@@ -0,0 +1,198 @@
+)
+
+const violationCategoryColorMap = {
+  blocker_violations: 'red',
+  critical_violations: 'orange',
+  major_violations: 'yellow',
+  minor_violations: 'yellowgreen',
+  info_violations: 'green',
+}
+
+module.exports = class SonarViolations extends SonarBase {
+  static get category() {
+    return 'analysis'
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'violations' }
+  }
+
+  static renderLongViolationsBadge(violations) {
+    if (violations.violations === 0) {
+      return {
+        message: 0,
+        color: 'brightgreen',
+      }
+    }
+
+    let color
+    const violationSummary = []
+
+    if (violations.info_violations > 0) {
+      violationSummary.push(`${violations.info_violations} info`)
+      color = violationCategoryColorMap.info_violations
+    }
+    if (violations.minor_violations > 0) {
+      violationSummary.unshift(`${violations.minor_violations} minor`)
+      color = violationCategoryColorMap.minor_violations
+    }
+    if (violations.major_violations > 0) {
+      violationSummary.unshift(`${violations.major_violations} major`)
+      color = violationCategoryColorMap.major_violations
+    }
+    if (violations.critical_violations > 0) {
+      violationSummary.unshift(`${violations.critical_violations} critical`)
+      color = violationCategoryColorMap.critical_violations
+    }
+    if (violations.blocker_violations > 0) {
+      violationSummary.unshift(`${violations.blocker_violations} blocker`)
+      color = violationCategoryColorMap.blocker_violations
+    }
+
+    return {
+      message: violationSummary.join(', '),
+      color,
+    }
+  }
+
+  static render({ violations, metricName, format }) {
+    if (metricName === 'violations') {
+      if (format === 'long') {
+        return this.renderLongViolationsBadge(violations)
+      }
+      return {
+        message: metric(violations),
+        color: violationsColorScale(violations),
+      }
+    }
+
+    const color =
+      violations === 0 ? 'brightgreen' : violationCategoryColorMap[metricName]
+
+    return {
+      label: getLabel({ metric: metricName }),
+      message: `${metric(violations)}`,
+      color,
+    }
+  }
+
+  static get route() {
+    return {
+      base: 'sonar',
+      pattern: `${patternBase}/:metric(violations|blocker_violations|critical_violations|major_violations|minor_violations|info_violations)`,
+      queryParamSchema: queryParamWithFormatSchema,
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Sonar Violations (short format)',
+        namedParams: {
+          protocol: 'https',
+          host: 'sonarcloud.io',
+          component: 'swellaby:azdo-shellcheck',
+          metric: 'violations',
+        },
+        queryParams: {
+          format: 'short',
+          version: '4.2',
+        },
+        staticPreview: this.render({
+          violations: 0,
+          metricName: 'violations',
+          format: 'short',
```

Comment:
Not sure what the best solution is for this - it might be a problem in the base class. Your examples work fine unless I try to change the params, so the "short" example works and the "long" example works, but if I click the "short" example and try to change it to "long" in the editor, it still shows the short version:

![Screenshot at 2019-04-13 19-38-08](https://user-images.githubusercontent.com/6025893/56083970-1a9f6c80-5e24-11e9-8ca5-aa8eb011e6df.png)

Reply:
Also, these examples generate a really really long example URL:

![Screenshot at 2019-04-13 19-41-58](https://user-images.githubusercontent.com/6025893/56083984-43bffd00-5e24-11e9-8dc6-c6250729039f.png)

Do you think it is worth aliasing the URL to just `:metric` in the examples given that all the many different options are discoverable via the drop-down in the editor?

Reply:
Lol, that's a pretty ugly url. Happy to change it.

Reply:
> Not sure what the best solution is for this - it might be a problem in the base class. Your examples work fine unless I try to change the params, so the "short" example works and the "long" example works, but if I click the "short" example and try to change it to "long" in the editor, it still shows the short version

I think that's because the specified component has zero violations. Both the long and short format render the same for zero violations, but if you use a different component that has violations you can toggle between them:

https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/violations.svg?format=long
![](https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/violations.svg?format=long)

https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/violations.svg?format=short
![](https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/violations.svg?format=short)

Reply:
You're right about the long/short param - I was just looking at a bad example :)

You're also right about shortening the URL example. The popup will use the route we define in the example definiteion. Lets leave it how it is and punt the job of thinking about how to deal with displaying (long) URLs to https://github.com/badges/shields/issues/3174